### PR TITLE
Feature/39520

### PR DESCRIPTION
**Pull request by sa11616079**
Create a new folder extraTs.
Define tsconfig.json.
Config is:

{
"compilerOptions": {
"module": "commonjs",
"target": "es6",
"noImplicitAny": false,
"moduleResolution": "node",
"allowSyntheticDefaultImports": true,
"resolveJsonModule": true,
"esModuleInterop": false,
"allowJs": true /* Allow javascript files to be compiled. */,
"checkJs": false /* Report errors in .js files. */,
"sourceMap": true,
"outDir": "./dist",
"baseUrl": ".",
"paths": {
"*": [
"node_modules/*",
"src/types/*"
]
},
},
"exclude": ["dist"],
"include": [
"extraTs"
],
"types": ["src/types"]
}

Install dependencies(as dev) for running typescript:
- typescript
- tsc-watch

Copy the folders (patterns and utils) and files (constants and index.js) from extra to extraTs.
Rename all files to .ts extension.
Also, create a file interfaces.ts to put all interfaces.
Now convert all the code to typescript as described in the class.

Add dist folder to .gitignore.

Acceptance Criteria:
- Every function must have types defined for arguments and also the return value.

Commands: 
 

"start:extra": "babel-node extra",

"start:extraTs": "tsc-watch --project . --onSuccess \"npm run nodemon\"",

"nodemon": "nodemon ./dist/index.js",

"test": "echo \"Error: no test specified\" && exit 1",

### CODE REVIEW

File: extra Ts/constants.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    'getUser2':
+    {
+        all: ['head-trainer'],
+        read: ['trainee', 'trainer'],
+        write: ['trainer'],
+        delete: []
+    }
+};
+
+import { Ipermissions , IUsers} from "./interfaces";
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/constants.ts:18]: import at the starting of file

File: extra Ts/pattern/Diamond.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
+function diamond(noOfRows)
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/pattern/Diamond.ts:1]: add argument type

File: extra Ts/pattern/Equilateral.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+function equilateral(noOfRows)
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/pattern/Equilateral.ts:1]: add argument type

File: extra Ts/pattern/index.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+import diamond from './Diamond.js';
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/pattern/index.ts:1]: the file extension is .ts

File: extra Ts/utils/index.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import {hasPermission} from './permissions.js';
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/utils/index.ts:1]: file extension is .ts

File: extra Ts/utils/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+//helper.js
+
+const users =
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/utils/helpers.ts:3]: keep your constants in constants.ts only

File: extra Ts/utils/permissions.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+function hasPermission(moduleName, role, permissionType) {
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/utils/permissions.ts:1]: add argument types

File: extra Ts/utils/validation.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
+let checkemail = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+let vldUser : String[]=[];
+let inVldUser : String[]=[];
+function validateEmail(email)
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/utils/validation.ts:4]: this should have been in helpers.ts with  argument and return types defined.

File: extra Ts/utils/validation.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+let checkemail = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+let vldUser : String[]=[];
+let inVldUser : String[]=[];
+function validateEmail(email) 
+{
+  return checkemail.test(String(email).toLowerCase());
+}
+
+function validateUsers(user)
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/utils/validation.ts:9]: add argument types

File: dist/constants.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+"use strict";
```
Comment (SHUBHAM-JAIN-SS) [inline dist/constants.js:1]: add dist folder to .gitignore

File: .editorconfig [hunk 11]
```diff
@@ -1,16 +0,0 @@
-root = true
```
Comment (SHUBHAM-JAIN-SS) [inline .editorconfig:1]: why have you deleted editorconfig?

File: extra Ts/constants.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+const permissions : Ipermissions={
```
Comment (SHUBHAM-JAIN-SS) [inline extra Ts/constants.ts:1]: the folder name seems to be having a space between extra and Ts, please remove it

**Review by SHUBHAM-JAIN-SS — CHANGES_REQUESTED** [review]
please make the required changes

**Comment by sa11616079** [comment]
resolved issues

**Pull request merged**